src/main/java/com/epam/aidial/core/service/ShareService.java

```diff
@@ -0,0 +1,257 @@
+                .collect(Collectors.groupingBy(ResourceLink::getResourceType));
+
+        for (Map.Entry<ResourceType, List<ResourceLink>> group : resourceGroups.entrySet()) {
+            ResourceType resourceType = group.getKey();
+            List<ResourceLink> links = group.getValue();
+            String ownerBucket = links.get(0).getBucket();
+            String ownerLocation = encryptionService.decrypt(ownerBucket);
+
+            ResourceDescription sharedByMe = getShareResource(ResourceType.SHARED_BY_ME, resourceType, ownerBucket, ownerLocation);
+            resourceService.computeResource(sharedByMe, state -> {
+                // if no state - set current
+
+                SharedByMeDto dto;
+                if (state == null) {
+                    Map<String, Set<String>> resourceToUsers = new HashMap<>();
+                    dto = new SharedByMeDto(resourceToUsers);
+                } else {
+                    dto = ProxyUtil.convertToObject(state, SharedByMeDto.class);
+                }
+
+                links.forEach(resourceLink -> {
+                    dto.addUserToResource(resourceLink.url(), location);
+                });
+
+                return ProxyUtil.convertToString(dto);
+            });
+
+            ResourceDescription sharedWithMe = getShareResource(ResourceType.SHARED_WITH_ME, resourceType, bucket, location);
+            resourceService.computeResource(sharedWithMe, state -> {
+                ResourceLinkCollection collection;
+
+                if (state == null) {
+                    collection = new ResourceLinkCollection(new HashSet<>());
+                } else {
+                    collection = ProxyUtil.convertToObject(state, ResourceLinkCollection.class);
+                }
+
+                collection.getResources().addAll(links);
+
+                return ProxyUtil.convertToString(collection);
+            });
+        }
+    }
+
+    public boolean hasReadAccess(String bucket, String location, ResourceDescription resource) {
+        ResourceDescription shareResource = getShareResource(ResourceType.SHARED_WITH_ME, resource.getType(), bucket, location);
+
+        String state = resourceService.getResource(shareResource);
+        if (state == null) {
+            log.info("No state found for share access");
+            return false;
+        } else {
+            ResourceLinkCollection sharedResources = ProxyUtil.convertToObject(state, ResourceLinkCollection.class);
+            // add parent folder search
+            return sharedResources.getResources().contains(new ResourceLink(resource.getUrl()));
+        }
+    }
+
+    public void revokeSharedAccess(String bucket, String location, ResourceLinkCollection resources) {
+        for (ResourceLink link : resources.getResources()) {
+            if (!link.getBucket().equals(bucket)) {
+                throw new IllegalArgumentException("You are only allowed to revoke access from own resources");
```

Comment:
Should not we write a warning message to the log and let process the rest of resources?